File: .tmp_tools/hr_sweep3.py
import asyncio, json, websockets

CDP = "ws://127.0.0.1:9340"
TARGET = "http://127.0.0.1:8922/"

async def main():
    async with websockets.connect(CDP, max_size=64*1024*1024) as ws:
        n = 0
        async def send(method, params=None):
            nonlocal n
            n += 1
            await ws.send(json.dumps({"id": n, "method": method, "params": params or {}}))
        # find a page target
        await send("Target.getTargets")
        while True:
            msg = json.loads(await ws.recv())
            if msg.get("id") == n:
                targets = msg["result"]["targetInfos"]
                pages = [t for t in targets if t["type"] == "page"]
                page = pages[0]
                if not pages: raise SystemExit("no page target")
                break
        sid = None
        await send("Target.attachToTarget", {"targetId": page["targetId"], "flatten": True})
        while True:
            msg = json.loads(await ws.recv())
            if msg.get("id") == n:
                sid = msg["result"]["sessionId"]
                break
        bad = []
        def recv_loop():
            try:
                while True:
                    msg = json.loads(ws.recv())
                    m = msg.get("method")
                    if m == "Network.responseReceived":
                        p = msg["params"]
                        if p["response"]["status"] >= 400:
                            bad.append((p["response"]["status"], p["response"].get("mimeType",""), p["response"]["url"][:140]))
            except Exception:
                pass
        import threading
        t = threading.Thread(target=recv_loop, daemon=True)
        t.start()
        async def s(method, params=None):
            nonlocal n
            n += 1
            await ws.send(json.dumps({"id": n, "sessionId": sid, "method": method, "params": params or {}}))
            while True:
                msg = json.loads(await ws.recv())
                if msg.get("id") == n:
                    return msg
        await s("Page.enable"); await s("Network.enable")
        await s("Page.navigate", {"url": TARGET})
        await asyncio.sleep(9)
        # scroll to bottom in waves
        await s("Runtime.evaluate", {"expression": "var i=0;function f(){window.scrollTo(0,document.body.scrollHeight);i++;"})
        for _ in range(6):
            await s("Runtime.evaluate", {"expression": "window.scrollTo(0,document.body.scrollHeight);"})
            await asyncio.sleep(1.5)
        await asyncio.sleep(4)
        # classify
        from collections import Counter
        c = Counter((st, m.split(';')[0] if m else '') for st, m, u in bad)
        print("== 404s ==", len(bad))
        for (st, mt), cnt in c.most_common():
            print(f"  {st} {mt} x{cnt}")
        for st, mt, u in sorted(set(bad)):
            print("   ", st, mt, u)
        # also tally all 4xx
        fourxx = [b for b in bad if b[0] >= 400]
        print("TOTAL >=400:", len(fourxx))

asyncio.run(main())

File: .tmp_tools/hr_sweep3.ts
const CDP = "ws://127.0.0.1:9340/devtools/browser/cf4517c1-747d-414f-9420-c0f5939c5b82";
const TARGET = "http://127.0.0.1:8922/";
let id = 0;
const pending = new Map<number, (m: any) => void>();
const errors: Array<[number, string, string]> = [];
const log = (...a: any[]) => console.log(new Date().toISOString().slice(11, 19), ...a);
const withTimeout = <T,>(p: Promise<T>, ms: number, what: string): Promise<T> =>
  Promise.race([p, new Promise<T>((_, rej) => setTimeout(() => rej(new Error("hang: " + what)), ms))]);
function send(method: string, params: any = {}, sessionId?: string, label = method) {
  return withTimeout(new Promise<any>((resolve, reject) => {
    const mid = ++id;
    pending.set(mid, (m: any) => (m.error ? reject(new Error(method + ": " + JSON.stringify(m.error).slice(0,200))) : resolve(m)));
    ws.send(JSON.stringify({ id: mid, sessionId, method, params }));
  }), 8000, label);
}
const ws = new WebSocket(CDP);
ws.addEventListener("open", () => log("open OK"));
ws.addEventListener("message", (ev: any) => {
  try {
    const msg = JSON.parse(ev.data);
    if (msg.id && pending.has(msg.id)) { pending.get(msg.id)!(msg); pending.delete(msg.id); }
    else if (msg.method === "Network.responseReceived") {
      const p = msg.params;
      if (p.response.status >= 400) errors.push([p.response.status, p.response.mimeType || "", p.response.url.slice(0, 150)]);
    }
  } catch (e) { log("parse err", String(e)); }
});
ws.addEventListener("error", (e: any) => log("WS error", JSON.stringify(e)));
await withTimeout(new Promise<void>((r) => ws.addEventListener("open", () => r())), 8000, "open");
const { result } = await send("Target.getTargets", {}, undefined, "getTargets");
log("targets", result.targetInfos.length);
const page = result.targetInfos.find((t: any) => t.type === "page");
if (!page) throw new Error("no page target");
log("using page", page.url.slice(0, 90));
const { result: att } = await send("Target.attachToTarget", { targetId: page.targetId, flatten: true }, undefined, "attach");
const sid = att.sessionId;
log("sid", sid);
await send("Page.enable", {}, sid, "Page.enable");
await send("Network.enable", {}, sid, "Network.enable");
await send("Page.navigate", { url: TARGET }, sid, "navigate");
log("navigated");
await withTimeout(new Promise<void>((r) => setTimeout(r, 9000)), 11000, "wait9");
for (let k = 0; k < 7; k++) {
  await send("Runtime.evaluate", { expression: "window.scrollTo(0, document.body.scrollHeight);" }, sid, "scroll");
  await withTimeout(new Promise<void>((r) => setTimeout(r, 1500)), 4000, "wait1.5");
}
await withTimeout(new Promise<void>((r) => setTimeout(r, 4000)), 6000, "wait4");
log("errors:", errors.length);
const uniq = new Map<string, [number, string, number]>();
for (const [st, mt, u] of errors) {
  const k = `${st}|${mt}|${u}`;
  const e = uniq.get(k) || [st, mt, 0];
  e[2] += 1; uniq.set(k, e);
}
for (const [st, mt, cnt] of uniq.values()) log("  " + st + " " + mt + " x" + cnt);
ws.close();
process.exit(0);
